### DISCUSSION TITLE
S101 Use of assert detected

### DISCUSSION THREAD

**Discussion by cidrblock** [ansible/ansible-navigator#713] [Linters, linting, formatting, pre-commit, CI] [answered]
This was in a test:

```
assert command == cmd_test_data.result_command, comment
assert additional_params == cmd_test_data.result_params, comment
```

this was a good discussion:

https://stackoverflow.com/questions/68428293/s101-use-of-assert-detected-for-python-tests

Maybe this should be disabled for the test directory with a link/explanation?

**Comment by webknjaz** [comment]
Even WPS itself disables this in tests. I asked them about it at some point, it's the official recommendation.

As for the main project codebase, it tends to be controversial. For example, ansible-core forbids it with an in-house linter requiring it to be rewritten as `raise AssertionError`. There are some valid use-cases for this but the problem is that an average Pythonista does not understand when it needs to be used which results in projects just blindly disallowing it globally (it's "all or nothing" because the validity totally depends on the human context and cannot be reliably guessed automatically).
That use-case is checking API contracts to state entry and exit expectations about a function's arguments and return values. There's also another popular use — hinting mypy about certain unclear type assumptions it may make.

Maybe we could disallow `assert` in `src/` but agree that for valid cases there should be a noqa+motivation comment.

**Comment by cidrblock** [comment, marked as answer]
this is covered by the following entry in the flake8 config

```
  # S101: Allow the use of assert within the tests directory, since tests require it.
  tests/**.py: S101
```
